crates/engine/src/prepare/operation/complexity_control.rs

```diff
@@ -0,0 +1,192 @@
+use engine_parser::types::OperationType;
+use schema::Schema;
+
+use crate::{
+    operation::{FieldWalker, OperationWalker, SelectionSetWalker},
+    prepare::error::{PrepareError, PrepareResult},
+};
+
+use super::Variables;
+
+pub fn control_complexity(schema: &Schema, operation: OperationWalker<'_>, variables: &Variables) -> PrepareResult<()> {
+    if schema.settings.complexity_control.is_disabled() {
+        return Ok(());
+    }
+
+    let base_cost = match operation.operation.ty {
+        OperationType::Query | OperationType::Subscription => 0,
+        OperationType::Mutation => 10,
+    };
+
+    let selection_set = operation.selection_set();
+
+    let context = ComplexityContext {
+        default_list_size: schema
+            .settings
+            .complexity_control
+            .list_size()
+            .expect("should be some unless disabled"),
+        variables,
+    };
+
+    let cost = base_cost + selection_set_complexity(&context, selection_set, None);
+
+    if let Some(limit) = schema.settings.complexity_control.limit() {
+        if cost > limit && schema.settings.complexity_control.is_enforce() {
+            return Err(PrepareError::ComplexityLimitReached);
+        }
+    }
+
+    Ok(())
+}
+
+struct ComplexityContext<'a> {
+    default_list_size: usize,
+    variables: &'a Variables,
+}
+
+fn selection_set_complexity(
```

Comment:
I'm a bit confused as to how you apply `sizedFields` to be honest. Maybe it makes sense, but feels like the traversing of the selection sets and the propagation of the estimated list size from the `@listSize` directive are too intertwined to make it clear what's happening to me. I don't find the enum `ChildCount` to be straightforward either.

Reply:
What are you confused about?

Reply:
Well, reading it it's not obvious to me if the logic is correctly applied. Maybe it's just a question of time, but I don't find `ChildCount` to be a very clear enum typically. For the sizeFields argumentsI would have expected something like this without an enum to propagate

```
fn compute_field_complexity(field, parent_count) {
  for field in selection_set {
     if field in sizeFields {
        compute_selection_set_complexity(field.subselection(), count)
     } else {
        compute_field_complexity(field, count)
     }
  }
  ...
}

fn compute_selection_set_complexity(selection_set, parent_count) {
  for field in selection_set {
        compute_field_complexity(field, count)
  }
  ...
}
```

But maybe I'm missing something 🤷